App Service cmd for importing certs from Key Vault

madsd opened the pull request:
App Service cmd for importing certs from Key Vault:
* az webapp|functionapp config ssl import

---

This checklist is used to make sure that common guidelines for a pull request are followed.

- [x] The PR has modified HISTORY.rst describing any customer-facing, functional changes. Note that this does not include changes only to help content. (see [Modifying change log](https://github.com/Azure/azure-cli/tree/master/doc/authoring_command_modules#modify-change-log)).

- [x] I adhere to the [Command Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/command_guidelines.md).

haroldrandom commented:
Please run `azdev linter {YOUR_MODULE}` to make sure command's help message is valid

madsd commented:
Hi @haroldrandom. I fixed the help error and some other styling. I also added a scenario test. The test works when I run it local, but fails in CI - the error is from creating an app service plan. The same lines of code found in numerous other tests. Any idea what is wrong?

haroldrandom commented:
@madsd Have you edited the recording YAML file manually?

yonzhan commented:
@jiasli  please help with the PR.

madsd commented:
No changes to the Yaml file, and I can run the tests in live mode on my machine without problems. The strange thing is, that it seem to fail on the basic appservice plan create. Don't know if there is a way to get more details on the exception.

panchagnula commented:
> No changes to the Yaml file, and I can run the tests in live mode on my machine without problems. The strange thing is, that it seem to fail on the basic appservice plan create. Don't know if there is a way to get more details on the exception.

you can delete the YAML file locally & re-record & upload the new yaml. You can also do is do a azdevsetup -c again on your local machine & see if you see this failure locally as well.

panchagnula commented on src/azure-cli/azure/cli/command_modules/appservice/commands.py:
```diff
@@ -143,6 +143,8 @@ def load_command_table(self, _):
         g.custom_command('unbind', 'unbind_ssl_cert', validator=validate_app_or_slot_exists_in_rg)
         g.custom_command('delete', 'delete_ssl_cert', exception_handler=ex_handler_factory())
+        g.custom_command('import', 'import_ssl_cert', exception_handler=ex_handler_factory(), is_preview=True)
+        g.custom_command('create', 'create_ssl_cert', exception_handler=ex_handler_factory(), is_preview=True)
```
Give we have AppService Certificates & Managed Certificates  this is targetting Free/Mamanged Certs only we should update this to name to reflect that this is for managed certificate by saying create-managed-certificate or something like that?

madsd replied:
Managed Certificate removed for now.

yungezz commented on src/azure-cli/azure/cli/command_modules/appservice/custom.py:
```diff
@@ -2005,6 +2005,52 @@ def delete_ssl_cert(cmd, resource_group_name, certificate_thumbprint):
+    kv_client = keyvault_client_vaults_factory(cmd.cli_ctx, None)
+    vault = kv_client.get(resource_group_name=resource_group_name, vault_name=key_vault_name)
+    # Check for Microsoft.Azure.WebSites app registration (app-id: abfa0a7c-a6b6-4736-8310-5855508787cd)
+    AZURE_PUBLIC_WEBSITES_APP_ID = 'abfa0a7c-a6b6-4736-8310-5855508787cd'
```
where to get these app Ids? is  it possible to import frorm dependencies?

madsd replied:
No, these are "well-known" guids and are documented in this official blog: https://azure.github.io/AppService/2016/05/24/Deploying-Azure-Web-App-Certificate-through-Key-Vault.html

yungezz commented:
is it necessary to support version of cert?

haroldrandom commented:
/azp run

madsd commented:
> is it necessary to support version of cert?

No, the commands setups up the integration with KV and pulls the latest version - any future versions of the certs are handled directly between App Service and Key Vault.

panchagnula commented on src/azure-cli/azure/cli/command_modules/appservice/custom.py:
```diff
@@ -2005,6 +2005,52 @@ def delete_ssl_cert(cmd, resource_group_name, certificate_thumbprint):
+
+    if not _check_service_principal_permissions(cmd, kv_resource_group_name, kv_name):
+        logger.warning('Unable to verify Key Vault permissions.')
+        logger.warning('You may need to grant Microsoft.Azure.WebSites service principal the Secret:Get permission')
```
Add link to the AppService blog here https://azure.github.io/AppService/2016/05/24/Deploying-Azure-Web-App-Certificate-through-Key-Vault.html

madsd replied:
Added link to warning (and renamed validate function to format per discussion)

madsd commented:
@yungezz, @haroldrandom requested changes has been implemented and the the latest History.rst conflict resolved.
